tempesta_fw/http_parser.c

```diff
@@ -1809,11 +1834,73 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
+        if (unlikely(IS_CR_OR_LF((c))))
+            __FSM_JMP(RGen_EoL);
+
+        if (unlikely(!IN_ALPHABET(c, hdr_a)))
+            return TFW_BLOCK;
+
+#ifdef __AVX2__
+        if (__data_available(p, 32)) {
+            __m256i text = _mm256_loadu(p);
+            AVX_LOWERCASE(text, text);
+            __m256i mask1, mask2;
+            //match operation is quite slow, and in most cases
+            //header name is 8-16 chars long, so we start 2 of them in parallel
+            //we will utilize later chars in order to fast-skip both header name,
+            //':', some spaces and part of header value if it is "other" type
+            AVX_MATCH_CHARSET(mask1, text, HEADER_CHARSET);
+            mask2 = _mm256_and_si256(
+                _mm256_cmpgt_epi8(text, _mm256_shuffle_epi32(__avx_constants2, __AVX_C2_CONTROL1)),
+                _mm256_cmpgt_epi8(_mm256_shuffle_epi32(__avx_constants2, __AVX_C2_CONTROL2), text));
+            //mask text for classification
+            int bitmask1 = _mm256_movemask_epi8(mask1);
+            int bitmask2 = _mm256_movemask_epi8(
+                _mm256_cmpeq_epi8(text, _mm256_shuffle_epi32(__avx_constants1, __AVX_C1_SEMICOLON)));
+            int bitmask3 = _mm256_movemask_epi8(
+                _mm256_cmpeq_epi8(text, _mm256_shuffle_epi32(__avx_constants1, __AVX_C1_SPACES)));
+            int bitmask4 = _mm256_movemask_epi8(mask2);
+            //at this point we must have something like this:
+            //           Host:   yandex.ru\n
+            //~bitmask1  ****    ****** **
+            // bitmask2      *
+            // bitmask3       ***
+            // bitmask4  *****************
+
+            //check for long header name
+            if (unlikely(!bitmask1))
+                return TFW_BLOCK;
+
+            int tmp = (~bitmask1)+1;
+            //bitmask2[__builtin_ffs(bitmask1)] must be 01
+            // tmp           *   ****** **
+            // bitmask2      *
+            tmp &= bitmask2;
+            if (unlikely(!tmp))
+                return TFW_BLOCK;
+
+            int headerid = 0;
+            mask1 = _mm256_andnot_si256(mask1, text);
+            AVX_MATCH_HEADER_NAME(headerid, mask1);
+            parser->_hdr_tag = AVX_STRID_0(headerid);
+            parser->_i_st    = AVX_STRID_1(headerid);
+            //now skip spaces after "header_name:"
+            // tmp           *
+            // bitmask3       *** xxxxxxx
+            tmp = tmp + tmp + bitmask3;
+            //clear all bits except first '1'
+            tmp = ((~tmp)+1) & tmp;
+            //1st '1' bit hits first header value byte
+            //           Host:   yandex.ru\n
+            // tmp               *
+            if (unlikely(!tmp)) {
+                //we got header name and some spaces after it, but
+                //not header value at all
+                __FSM_MOVE_n(RGen_LWS, 32);
+            }
+            __FSM_MOVE_n(RGen_LWS, __builtin_ctz(tmp));
+        }
+#endif
```

Comment:
This is not subject for #182. Meantime, `memchr()` which ARE subject of #182 still exist in code, so 'other' headers aren't verified for allowed alphabet.